### PR TITLE
View group messages

### PR DESCRIPTION
**Pull request by Mukalawit**
**What does this PR do?**
This PR enables users to view messages in the broadcast groups they belong to in to the PostIt Application.

**Description of the task to be completed**
For registered group users to be able to add view messages to their broadcast group(s)

**Why do we need this feature?**
To enable users to view messages in broadcast groups they belong to

**How can this feature be manually tested?**
Using **Postman** GET:` http:localhost:3000/api/group/:id/messages`

**Testing** 

```
npm run test

```

### CODE REVIEW

File: server/routes/message/view-group-message.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const { Message } = require('../../models');
+const { authenticateToken } = require('../../utils/token-helpers');
+
+const router = express.Router();
+router.get('/:id/messages', authenticateToken, async (req, res) => {
+    const { id } = req.user;
+    const groupId  = Number(req.params.id);
+
+    try {
+        const message = new Message(id,groupId);
```
Comment (Descartess) [inline server/routes/message/view-group-message.js:12]: I  don't think it's a good design to create a dummy instance of Message so that we can fetch group messages. 

The method to get group messages should be a static method on the Message class and it should be used as shown below. 

```suggestion
        const messages = Message.fetchGroupMessages(id,groupId);
```

File: server/routes/message/view-group-message.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        await message.view();
+        return res.status(200).json({ message: 'Viewing Messages' });
```
Comment (Descartess) [inline server/routes/message/view-group-message.js]: This should return the actual messages for the group
```suggestion
        return res.status(200).json({ message:  messages });
```

File: server/models/messages.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 const {HttpError} = require('../utils/errors');
 const prisma = require('../db');
 class Message {
- constructor(userId,message,groupId){
+ constructor(userId,groupId,message=""){
```
Comment (Descartess) [inline server/models/messages.js:4]: This opens up a scenario where messages are just empty strings. I don't see any value in having that. I think we should revert this change. 

I believe this change was added so that you could create a new instance of message to fetch other group messages.

File: server/models/messages.js [hunk 4]
```diff
@@ -28,6 +28,17 @@ class Message {
       }
    });
  }
+ async view(){
```
Comment (Descartess) [inline server/models/messages.js:31]: We can make this a static method and even rename it. This current design is forcing us to create objects of dummy messages so that we can get the real messages. 

I have written a suggestion below.

File: server/routes/message/post-group-message.js [hunk 5]
```diff
@@ -13,7 +13,7 @@ router.post('/:id/message', authenticateToken, async (req, res) => {
     }
 
     try {
-        const message = new Message(id,msg,groupId);
+        const message = new Message(id,groupId,msg);
```
Comment (Descartess) [inline server/routes/message/post-group-message.js]: Let's try to stay away from abbreviations.  `msg` could be renamed to `text` or `content`. 
```suggestion
        const message = new Message(id,groupId,text);
```

File: server/__tests__/view-group-messages-tests.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+const baseURL = 'http://localhost:3001';
+
+describe('When a user views messages in a group', () => {
+  it('should return correct status code if the user is viewing messages', async () => {
+    Message.mockImplementationOnce(() => ({
+      view: jest.fn(() => Promise.resolve()),
+    }));
+    const response = await request(baseURL).get('/api/group/:id/messages').set('Authorization', `Bearer ${token}`);
+    expect(response.statusCode).toBe(200);
```
Comment (Descartess) [inline server/__tests__/view-group-messages-tests.js:42]: You can also assert the `response.body` to return the actual messages.  

You enforce that the endpoint returns the real messages as opposed to `Viewing messages`

Comment (Descartess) [inline server/__tests__/view-group-messages-tests.js:42]: Here's some pseudo code for the your reference
```suggestion
    expect(response.statusCode).toBe(200);
     expect(response.data.messages).toBe([message]);
```

**Review by Descartess — CHANGES_REQUESTED** [review]

**Comment by Descartess** [comment]
Good first attempt.

File: server/routes/message/view-group-message.js (hunk 1 above)
Comment (Mukalawit) [reply server/routes/message/view-group-message.js:12]: It cannot be a static method because it uses `this` keyword in its internal implementation

File: server/routes/message/view-group-message.js (hunk 2 above)
Comment (Mukalawit) [reply server/routes/message/view-group-message.js]: Thanks

File: server/models/messages.js (hunk 3 above)
Comment (Mukalawit) [reply server/models/messages.js:4]: The implementation of having an optional message argument for the constructor  in this class , gives us the flexibility of using the view function without  any unecessary modifications to the class(keeping it DRY),  and also message input validation is part of the route implementation for post group messages, so we never have to worry if the message argument has been provided to the constructor or if it will ever be empty under any other scenario.

File: server/models/messages.js (hunk 4 above)
Comment (Mukalawit) [reply server/models/messages.js:31]: Please refer to my earlier reasons as to why it cannot be static

File: server/routes/message/view-group-message.js (hunk 1 above)
Comment (Descartess) [reply server/routes/message/view-group-message.js:12]: You can refactor the internal implementation not to use this keyword and have the dependencies passed down as variables

**Review by Descartess — APPROVED** [review]
